docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,287 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
+> One option is to put all these methods under `Json.as_xxx()`,
+> so `Json.as_num(Json 1234) == 1234`. But this won't work very well for user-defined
+> structs...
+>
+> ALSO: We should rename `str`/`num`/`bool` to `Str`/`Num`/`Bool`. Thoughts?
+>
+
+All `parse_json()` methods will validate that the runtime type is compatible with the target type in
+order to ensure type safety (at a runtime cost):
+
+```js
+str.parse_json(json_number);      // RUNTIME ERROR: unable to parse number `123` as a string.
+num.parse_json(Json "\"hello\""); // RUNTIME ERROR: unable to parse string "hello" as a number
+
+let my_array = Json [1,2,3,"hello"];
+Array<num>.parse_json(my_array); // RUNTIME ERROR: unable to parse `[1,2,3,"hello"]` as an array of `num`.
+```
+
+Use `unsafe: true` to disable this check at your own risk:
+
+```js
+let x = num.parse_json(Json [1,2,3], unsafe: true);
+print(x + 5); // OUTPUT: 1,2,35
+```
+
+For each `parse_json()`, there is a `try_parse_json()` method which returns an optional `T?` which
+indicates if parsing was successful or not:
+
+```js
+let s = str.try_parse_json(my_json) ?? "invalid string";
+```
+
+##### 1.1.4.2. Structs and Schemas
+
+All [structs](#31-structs) also have a `parse()` method that can be used to parse `Json` into a
+struct:
+
+```js
+struct Contact {
+  first: str;
+  last: str;
+  phone: str?;
+}
+
+let j = Json { first: "Wing", last: "Lyly" };
+let my_contact = Contact.parse(j);
+assert(my_contact.first == "Wing");
+```
+
+When a `Json` is parsed into a struct, the schema will be validated to ensure the result is
+type-safe:
+
+```js
+let p = Json { first: "Wing", phone: 1234 };
+Contact.parse(p);
+// RUNTIME ERROR: unable to parse Contact:
+// - field "last" is required and missing
+// - field "phone" is expected to be a string, got number.
+```
+
+Same as with primitives and containers, it is possible to opt-out of validation using `unsafe:
+true`:
+
+```js
+let p = Json { first: "Wing", phone: 1234 };
+let x = Contact.parse(p, unsafe: true);
+assert(x.last.len > 0);
+// RUNTIME ERROR: Cannot read properties of undefined (reading 'length')
+```
+
+Struct parsing is *partial* by default. This means that parsing is successful even if the `Json`
+includes extraneous fields:
+
+```js
+let p = Json { first: "hello", last: "world", another_field: "ignored" };
+let c = Contact.parse(p);
+assert(c.first == "hello");
+assert(c.last == "world");
+// `c.another_field` is not a thing
+```
+
+This can be disabled using `partial: false` (P2):
+
+```js
+Contact.parse(Json { first: "hello", last: "world", another_field: "ignored" }, partial: false);
+// RUNTIME ERROR: cannot parse Contact due to extraneous field "another_field"
+```
+
+Structs have a `schema` static method which returns a `JsonSchema` object (P2):
+
+```js
+let schema = Contact.schema();
+schema.validate(j);
+```
+
+##### 1.1.4.3. MutJson
+
+To define a mutable JSON container, use the `MutJson` type:
+
+```js
+let my_obj = MutJson { hello: "dear" };
```

Comment:
Are there any mutable operations you can do on a `MutJson "hello"` or `MutJson true`?

Reply:
Not really, but for the sake of completeness, you can still define it like that.

Reply:
added a clarification